onvm/onvm_nflib/onvm_nflib.c

```diff
@@ -1224,39 +1237,50 @@ onvm_nflib_terminate_children(struct onvm_nf_info *nf_info) {
                                continue;
 
                         /* Wake up the child if its sleeping */
-                        if (ONVM_ENABLE_SHARED_CPU && rte_atomic16_read(nfs[i].sleep_state) == 1) {
-                                rte_atomic16_set(nfs[i].sleep_state, 0);
-                                sem_post(nfs[i].nf_mutex);
+                        if (ONVM_ENABLE_SHARED_CPU && rte_atomic16_read(nfs[i].shared_core.sleep_state) == 1) {
+                                rte_atomic16_set(nfs[i].shared_core.sleep_state, 0);
+                                sem_post(nfs[i].shared_core.nf_mutex);
                         }
                 }
                 RTE_LOG(INFO, APP, "NF %d: Waiting for %d children to exit\n",
-                        nf_info->instance_id, rte_atomic16_read(&nfs[nf_info->instance_id].children_cnt));
+                        nf->instance_id, rte_atomic16_read(&nf->thread_info.children_cnt));
                 sleep(NF_TERM_WAIT_TIME);
                 iter_cnt++;
         }
 
-        if (rte_atomic16_read(&nfs[nf_info->instance_id].children_cnt) > 0) {
+        if (rte_atomic16_read(&nf->thread_info.children_cnt) > 0) {
                 RTE_LOG(INFO, APP, "NF %d: Up to %d children may still be running and must be killed manually\n",
-                        nf_info->instance_id, rte_atomic16_read(&nfs[nf_info->instance_id].children_cnt));
+                        nf->instance_id, rte_atomic16_read(&nf->thread_info.children_cnt));
         }
 }
 
 static void
-onvm_nflib_cleanup(struct onvm_nf_context *nf_context) {
+onvm_nflib_cleanup(struct onvm_nf_local_ctx *nf_local_ctx) {
```

Comment:
double check if you're freeing all of the pointers in the `onvm_nf` and `onvm_nf_local_ctx` struct.  this gets really weird because of the circular reference that i mentioned before.  you have to store `nf_local_ctx->nf` in a local ptr, then decide which struct needs to be freed f first as they both have references to each other, check each ptr in each struct to see if it is not null and then free it.

i see that some ptrs are not cleaned here.  also, i don't see the circular reference being handled correctly either.

Reply:
Yeah I actually had a patch for that but instead of a new PR I'll just apply it here.

Reply:
Should be resolved now